File: Problems/NZOI/NZIC Practice Problems/Python/Battleships.py
# https://train.nzoi.org.nz/problems/779

values = {
    "a": 1,
    "b": 2,
    "c": 3,
    "d": 4,
    "e": 5,
    "f": 6
}

grid = []

for i in range(10):
    arr = list(input().replace(" ", ""))
    grid.append(arr)

x, y = map(int, input().split())

while x != -1 and y != -1:
    target = grid[y][x]

    if target != "#" and values[target] == 1:
        print(f"Sunk {target}")
        del values[target]
        grid[y][x] = "#"


    elif target in values:
        print(f"Hit {target}")
        values[target] -= 1
        grid[y][x] = "#"
    
    else:
        print("Miss")

    x, y = map(int, input().split())

File: Problems/NZOI/NZIC Practice Problems/C++/Battleships.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    unordered_map<string, int> ships = {
        {"a", 1}, 
        {"b", 2}, 
        {"c", 3}, 
        {"d", 4}, 
        {"e", 5}, 
        {"f", 6}
    };

    unordered_map<string, int> health = ships;

    const int size = 10;
    vector<vector<string>> grid(size, vector<string>(size));

    for (int i = 0; i < size; i++) 
    {
        for (int j = 0; j < size; j++) 
        {
            cin >> grid[i][j];
        }
    }
    
    int x, y;

    while (cin >> x >> y && (x != -1 || y != -1))
    {
        string place = grid[y][x];

        if (place != "#") 
        {
            health[place]--;

            if (health[place] <= 0) 
            {
                cout << "Sunk " << place << "\n";
            } 

            else 
            {
                cout << "Hit " << place << "\n";
            }

            grid[y][x] = "#";

        } 

        else 
        {
            cout << "Miss" << "\n";
        }
    }
        
   return 0;
}